src/rulesets/Base/base.jl

```diff
@@ -88,3 +88,55 @@ function rrule(::typeof(identity), x::Tuple)
     end
     return x, identity_pullback
 end
+
+#####
+##### `evalpoly`
+#####
+
+if VERSION ≥ v"1.4"
+    function frule((_, Δx, Δp), ::typeof(evalpoly), x, p)
+        q = _evalpoly_dxcoef(p)
+        return evalpoly(x, p), evalpoly(x, Δp) + evalpoly(x, q) * Δx
+    end
+
+    function rrule(::typeof(evalpoly), x, p)
+        function evalpoly_pullback(Δy)
+            ∂x = Thunk() do
+                q = _evalpoly_dxcoef(p)
+                return evalpoly(x, q)' * Δy
+            end
+            ∂p = @thunk _evalpoly_backp(Δy, x, p)
+            return NO_FIELDS, ∂x, ∂p
+        end
+        return evalpoly(x, p), evalpoly_pullback
+    end
+
+    _evalpoly_dxcoef(p) = ntuple(i -> i * p[i + 1], length(p) - 1)
+    function _evalpoly_dxcoef(p::AbstractVector)
+        N = length(p)
+        @inbounds q = (1:(N - 1)) .* view(p, 1:(N - 1))
+        return q
+    end
+
+    # This is a geometric progression, that is ∂p = Δy * (I, x, x², …, xⁿ)'
+    # TODO: Handle case where x is a matrix and p is UniformScaling
+    function _evalpoly_backp(Δy, x, p::Tuple)
+        N = length(p)
+        ∂p = ntuple(_ -> Δy, N)
+        x′ = x'
+        for i in 2:N
+            ∂p = Base.setindex(∂p, ∂p[i - 1] * x′, i)
```

Comment:
TIL `Base.setindex` on `Tuple`

Reply:
Is this actually faster than
either:
```julia
ntuple(N) do i
    Δy*x′^i
end
```
or
```julia
last_val = Δy
tt = ntuple(N-1) do i
    last_val*=x′
end
∂p = (Δy, tt...)
```

or
```julia
last_val = Δy
∂p = ntuple(N) do i
    i==1 && return  Δy 
    last_val*=x′
end
```

Reply:
Quite a bit:
```julia
    function _evalpoly_backp(Δy, x, p::Tuple)
        N = length(p)
        ∂p = ntuple(_ -> Δy, N)
        x′ = x'
        for i in 2:N
            ∂p = Base.setindex(∂p, ∂p[i - 1] * x′, i)
        end
        return ∂p
    end
    function _evalpoly_backp2(Δy, x, p::Tuple)
        N = length(p)
        x′ = x'
        ∂p = ntuple(N) do i
            Δy * x′^(i - 1)
        end
        return ∂p
    end
    function _evalpoly_backp3(Δy, x, p::Tuple)
        N = length(p)
        x′ = x'
        last_val = Δy
        tt = ntuple(N - 1) do i
            last_val *= x′
        end
        ∂p = (Δy, tt...)
        return ∂p
    end
    function _evalpoly_backp4(Δy, x, p::Tuple)
        N = length(p)
        x′ = x'
        last_val = Δy
        ∂p = ntuple(N) do i
            i == 1 && return Δy 
            return last_val *= x′
        end
        return ∂p
    end
```

```julia
julia> using BenchmarkTools

julia> Δy, x, p = randn(), randn(), Tuple(randn(10));

julia> @btime _evalpoly_backp(Δy, x, p);
  53.628 ns (1 allocation: 96 bytes)

julia> @btime _evalpoly_backp2(Δy, x, p);
  261.738 ns (1 allocation: 96 bytes)

julia> @btime _evalpoly_backp3(Δy, x, p);
  1.356 μs (34 allocations: 704 bytes)

julia> @btime _evalpoly_backp4(Δy, x, p);
  745.799 ns (23 allocations: 464 bytes)
```

Reply:
Nice, I am going to start using this

Reply:
@oxinabox Actually I realized I was just lucky here that I picked `length(p) == 10`. if `length(p) > 10`, the version with `setindex` has _terrible_ performance.